File: week 6/sentimental-credit/credit.py
import cs50
from math import *
import sys

max = 16
key = 0
digits = []
credit = cs50.get_int("Number: ")
len = ceil(log(credit)/log(10))
temp = credit

if len > max:
    print("INVALID")
    sys.exit(1)

while (True):
    digits.append(int(temp % 10))
    temp /= 10
    if (temp <= 0):
        break

for i in range(len):
    if i % 2 == 1:
        if digits[i] > 4:
            key += ((2 * digits[i] % 100 - (2 * digits[i]) % 10) / 10) + ((2 * digits[i]) % 10)
        else:
            key += 2 * digits[i]
    else:
        key += digits[i]

if key % 10 != 0:
    print("INVALID")
    sys.exit(2)

if ((len == 13 or len == 16) and digits[len - 1] == 4):
    print("VISA")
elif (len == 16 and digits[15] == 5 and digits[14] in range(1, 6)):
    print("MASTERCARD")
elif (len == 15 and digits[14] == 3 and (digits[13] == 4 or digits[13] == 7)):
    print("AMEX")
else:
    print("INVALID")
    sys.exit(3)

File: C/week 1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int i;
int a = 0;
long k;
int q[20];

void digit(int p, long c);

int main(void)
{
    long credit = get_long("Number: ");
    for (i = 20; i > 0; i--)
    {
        int c = ((credit % (long) pow(10, i)) - (credit % (long) pow(10, i - 1))) /
                (long) pow(10, i - 1);
        if (c != 0)
        {
            k = i;
            i = -777;
        }
    }
    digit(k, credit);

    if (a % 10 == 0)
    {
        if (q[k] == 4 && (k == 13 || k == 16))
        {
            printf("VISA\n");
        }
        else if (q[k] == 5 &&
                 (q[k - 1] == 1 || q[k - 1] == 2 || q[k - 1] == 3 || q[k - 1] == 4 ||
                  q[k - 1] == 5) &&
                 k == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (q[k] == 3 && (q[k - 1] == 4 || q[k - 1] == 7) && k == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

void digit(int p, long c)
{
    for (int z = 1; z <= p; z++)
    {
        q[z] = ((c % (long) pow(10, z)) - (c % (long) pow(10, z - 1))) / (long) pow(10, z - 1);
        if (z % 2 == 0)
        {
            if (2 * q[z] >= 10)
            {
                a += ((2 * q[z] % 100 - (2 * q[z]) % 10) / 10) + ((2 * q[z]) % 10);
            }
            else
            {
                a += 2 * q[z];
            }
        }
        else
        {
            a += q[z];
        }
    }
}
